scala/src/com/google/idea/blaze/scala/run/producers/ScalaTestContextProvider.java

```diff
@@ -76,32 +92,83 @@ private static ScClass getTestClass(ConfigurationContext context) {
       // handled by a different producer
       return null;
     }
-    return getTestClass(location);
+    ScalaTestConfigurationProducer scalaTestConfigurationProducer = ScalaTestConfigurationProducer.instance();
+    return scalaTestConfigurationProducer.getTestClassWithTestName(location)
+        .map(classWithTestName -> new ClassAndTestName(classWithTestName.testClass(), classWithTestName.testName().getOrElse(() -> null)))
+        .getOrElse(() -> null);
   }
 
-  @Nullable
-  private static ScClass getTestClass(Location<?> location) {
-    PsiElement element = location.getPsiElement();
-    ScClass testClass;
-    if (element instanceof ScClass) {
-      testClass = (ScClass) element;
-    } else {
-      testClass = PsiTreeUtil.getParentOfType(element, ScClass.class);
+  public static class ScalatestTestSelectorFlagsModification implements TestContext.BlazeFlagsModification {
+    @Nullable
+    private final String testName;
+
+    private final String testClassSelectorFlag;
+    private final String testClassFlag;
+    @Nullable
+    private final String testNameSelectorFlag;
+    @Nullable
+    private final String testNameFlag;
+
+    public ScalatestTestSelectorFlagsModification(String testClassFqn, @Nullable String testName) {
+      this.testName = testName;
+      testClassSelectorFlag = BlazeFlags.TEST_ARG + "-s";
+      testClassFlag = BlazeFlags.TEST_ARG + testClassFqn;
+      testNameSelectorFlag = BlazeFlags.TEST_ARG + "-t";
+      if (testName != null) {
+        // Scalatest names can contain spaces, so the name needs to be quoted
+        // This means we need to escape " in the test name
+        String escapedTestName = testName.replace("\"", "\\\"");
+        testNameFlag = BlazeFlags.TEST_ARG + "\"" + escapedTestName + "\"";
+      } else {
+        testNameFlag = null;
+      }
     }
-    if (testClass != null && isTestClass(testClass)) {
-      return testClass;
+
+    private boolean hasTestName() {
+      return testName != null;
     }
-    return null;
-  }
 
-  private static boolean isTestClass(ScClass testClass) {
-    TestFramework framework = TestFrameworks.detectFramework(testClass);
-    return framework instanceof ScalaTestTestFramework && framework.isTestClass(testClass);
-  }
+    private boolean flagsMatchSettings(List<String> flags) {
+      int classSelectorIndex = flags.indexOf(testClassSelectorFlag);
+      if (classSelectorIndex == -1) {
+        return false;
+      }
+      int expectedSize = hasTestName() ? 4 : 2;
+      if (flags.size() < classSelectorIndex + expectedSize) {
+        return false;
+      }
+      String classSelector = flags.get(classSelectorIndex);
+      String clazz = flags.get(classSelectorIndex + 1);
+      boolean matchesClassFlags = classSelector.equals(testClassSelectorFlag) && clazz.equals(testClassFlag);
+      if (!hasTestName()) {
+        boolean hasTestSelector = flags.contains(testNameSelectorFlag);
```

Comment:
Based on that `testNameSelectorFlag` is set in line 116 before checking that testName exists or not. this check can cause the method to return false if flags list has `testNameSelectorFlag` but no `testName`.

Reply:
If flags contains `testNameSelectorFlag` but no `testName`, I think we want this method to return false in this case? What do you see as the issue here?

Reply:
Resolving this, since I think the behavior is correct